### PR TITLE
OtsuSplit WIP

### LINKED ISSUE #26: Port Python OtsuSplit

### PR DESCRIPTION
**Pull request by anlava**
Resolve #26 

- [x] Add threshold method

### CODE REVIEW

File: src/features/otsu_split.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        let mut dm = T::zero();
+        let mut w: usize = 0;
+        let mean = ts.m.get_mean();
+        let count = ts.lenu();
+        let msorted = ts.m.get_sorted();
+        let mut last_variance = T::zero();
+
+        for &m in msorted.iter() {
```
Comment (hombit) [inline src/features/otsu_split.rs:88]: We can use `.enumerate()` or `(1..).zip(msorted.iter())` instead, in this case we don't need a mutable `w`

File: src/features/otsu_split.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            if variance < last_variance {
+                break;
+            }
+            last_variance = variance;
+        }
+
+        let std_lower;
+        let std_upper;
+        let mean_lower;
+        let mean_upper;
```
Comment (hombit) [inline src/features/otsu_split.rs:108]: I believe we can try to find these values iteratively in the for-loop. `mean_upper - mean_lower` is just `dm`, `std_lower` should be faster to find in the for-loop, and `std_upper` is probably a function of `std_lower`, `dm` and `get_nonzero_m_std2(ts)?`.

File: src/features/otsu_split.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        let mean = ts.m.get_mean();
+        let count = ts.lenu();
+        let msorted = ts.m.get_sorted();
+        let mut last_variance = T::zero();
+
+        for &m in msorted.iter() {
+            w += 1;
+            dm += mean - m;
+
+            if w == count {
```
Comment (hombit) [inline src/features/otsu_split.rs:92]: It looks like it is possible only if time series is flat, but I could be wrong

File: src/features/otsu_split.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        // let mean_diff = upper.get_mean() - lower.get_mean();
+        let mean_diff = mean_upper - mean_lower;
+        let lower_to_all = (w - 1).value_as::<T>().unwrap() / count.value_as::<T>().unwrap();
+
+        Ok(vec![mean_diff, std_lower, std_upper, lower_to_all])
+    }
+}
+
+#[cfg(test)]
```
Comment (hombit) [inline src/features/otsu_split.rs:136]: Please test more edge-cases: 1) minimum number of observations, 2) `w == count` (if you could find such an example for non-flat time-series), 3) w == 2

**Review by hombit — CHANGES_REQUESTED** [review]

File: src/features/otsu_split.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    T: Float,
+{
+    fn eval(&self, ts: &mut TimeSeries<T>) -> Result<Vec<T>, EvaluatorError> {
+        self.check_ts_length(ts)?;
+
+        if ts.m.get_max() == ts.m.get_min() {
+            return Err(EvaluatorError::FlatTimeSeries);
+        }
+
+        let mut dm = T::zero();
```
Comment (hombit) [inline src/features/otsu_split.rs:81]: Maybe rename to `delta_mean`?

**Comment by hombit** [comment]
1. Docs: say about possible multiple maxima
2. Add benchmark here: https://github.com/light-curve/light-curve-feature/blob/e8b54b0a60b0259b2382093d5ab56f923287ab0d/benches/extractor.rs#L17
`cargo bench -- FeatureExtractor`

**Comment by anlava** [comment]
Benchmark results:

**FeatureExtractor otsu_mean_diff: [100; f64]                                                                            
                        time:   [264.12 ns 264.32 ns 264.55 ns]**
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe

**FeatureExtractor otsu_mean_diff: [1000; f64]                                                                             
                        time:   [1.8969 us 1.8988 us 1.9010 us]**
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

**Review by hombit — APPROVED** [review]
